pkg/query-service/app/inframetrics/deployments.go

```diff
@@ -0,0 +1,444 @@
+		return nil, nil, nil
+	}
+
+	if req.OrderBy.Order == v3.DirectionDesc {
+		sort.Slice(formattedResponse[0].Series, func(i, j int) bool {
+			return formattedResponse[0].Series[i].Points[0].Value > formattedResponse[0].Series[j].Points[0].Value
+		})
+	} else {
+		sort.Slice(formattedResponse[0].Series, func(i, j int) bool {
+			return formattedResponse[0].Series[i].Points[0].Value < formattedResponse[0].Series[j].Points[0].Value
+		})
+	}
+
+	limit := math.Min(float64(req.Offset+req.Limit), float64(len(formattedResponse[0].Series)))
+
+	paginatedTopDeploymentGroupsSeries := formattedResponse[0].Series[req.Offset:int(limit)]
+
+	topDeploymentGroups := []map[string]string{}
+	for _, series := range paginatedTopDeploymentGroupsSeries {
+		topDeploymentGroups = append(topDeploymentGroups, series.Labels)
+	}
+	allDeploymentGroups := []map[string]string{}
+	for _, series := range formattedResponse[0].Series {
+		allDeploymentGroups = append(allDeploymentGroups, series.Labels)
+	}
+
+	return topDeploymentGroups, allDeploymentGroups, nil
+}
+
+func (d *DeploymentsRepo) GetDeploymentList(ctx context.Context, req model.DeploymentListRequest) (model.DeploymentListResponse, error) {
+	resp := model.DeploymentListResponse{}
+
+	if req.Limit == 0 {
+		req.Limit = 10
+	}
+
+	if req.OrderBy == nil {
+		req.OrderBy = &v3.OrderBy{ColumnName: "cpu", Order: v3.DirectionDesc}
+	}
+
+	if req.GroupBy == nil {
+		req.GroupBy = []v3.AttributeKey{{Key: k8sDeploymentNameAttrKey}}
+		resp.Type = model.ResponseTypeList
+	} else {
+		resp.Type = model.ResponseTypeGroupedList
+	}
+
+	step := int64(math.Max(float64(common.MinAllowedStepInterval(req.Start, req.End)), 60))
+
+	query := WorkloadTableListQuery.Clone()
+
+	query.Start = req.Start
+	query.End = req.End
+	query.Step = step
+
+	// add additional queries for deployments
+	for _, deploymentQuery := range builderQueriesForDeployments {
+		query.CompositeQuery.BuilderQueries[deploymentQuery.QueryName] = deploymentQuery
+	}
+
+	for _, query := range query.CompositeQuery.BuilderQueries {
+		query.StepInterval = step
+		if req.Filters != nil && len(req.Filters.Items) > 0 {
+			if query.Filters == nil {
+				query.Filters = &v3.FilterSet{Operator: "AND", Items: []v3.FilterItem{}}
+			}
+			query.Filters.Items = append(query.Filters.Items, req.Filters.Items...)
+		}
+		query.GroupBy = req.GroupBy
+		// make sure we only get records for deployments
+		query.Filters.Items = append(query.Filters.Items, v3.FilterItem{
+			Key:      v3.AttributeKey{Key: k8sDeploymentNameAttrKey},
+			Operator: v3.FilterOperatorExists,
+		})
+	}
+
+	deploymentAttrs, err := d.getMetadataAttributes(ctx, req)
+	if err != nil {
+		return resp, err
+	}
+
+	topDeploymentGroups, allDeploymentGroups, err := d.getTopDeploymentGroups(ctx, req, query)
+	if err != nil {
+		return resp, err
+	}
+
+	groupFilters := map[string][]string{}
+	for _, topDeploymentGroup := range topDeploymentGroups {
+		for k, v := range topDeploymentGroup {
```

Comment:
nit: k, v -> actual key value pairs for readability

Reply:
in this case, the name of the key is not known ahead because it will be group by from the request. I can rename to `groupKey` but i am not sure if it improves readability.

Reply:
okay, lgtm